### PR TITLE
Allow logging of metrics in both validation_step and validation_epoch_end

### LINKED ISSUE #3215: validation_epoch_end not logging validation_step EvalResult values
## 🐛 Bug

When overwriting `validation_epoch_end` the EvalResult values from `validation_step` are not logged.

For my experiments I keep track of several metrics that I only log to TensorBoard at the end of each validation epoch. For most metrics I can specify `EvalResult().log(on_epoch=True)`, but one of the metrics I can only calculate at the end of the epoch. If I use `validation_epoch_end` to calculate this metric, the results from `validation_step` are not logged. 

### To Reproduce
```
    def validation_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)

        result = pl.EvalResult(checkpoint_on=loss)
        result.log('val/loss', loss, on_step=False, on_epoch=True)
        result.log('val/y_hat', y_hat, logger=False, on_step=False, on_epoch=False)
        return result

    def validation_epoch_end(self, outputs):
        y_hat = outputs['val/y_hat']
        example = y_hat.exp().mean()

        result = pl.EvalResult()
        result.log('val/epoch_end_metric', example)
        return result
```
The code above writes 'val/epoch_end_metric' to TensorBoard, but does nothing with the result from `validation_step`. 

As a result, the checkpoint metric specified in `validation_step` is also lost. Warnings are printed to screen, but it still prints a statement at the end of fit claiming to save the checkpoint (which is not happening).

```
...

  | Name | Type   | Params
--------------------------------
0 | l1   | Linear | 7 K   

Epoch 0:  50%|██████▌      | 5/10 [00:00<00:00, 549.90it/s, loss=2.454, v_num=2]
Validating: 0it [00:00, ?it/s] ---/pytorch_lightning/utilities/distributed.py:37: RuntimeWarning: The metric you returned None must be a `torch.Tensor` instance, checkpoint not saved HINT: what is the value of loss in validation_epoch_end()?
  warnings.warn(*args, **kwargs)
---/pytorch_lightning/utilities/distributed.py:37: RuntimeWarning: Can save best model only with loss available, skipping.
  warnings.warn(*args, **kwargs)

...

Epoch 4: 100%|████████████| 10/10 [00:00<00:00, 911.31it/s, loss=1.847, v_num=2]
                              Saving latest checkpoint..
Epoch 4: 100%|████████████| 10/10 [00:00<00:00, 761.27it/s, loss=1.847, v_num=2]
```

### Expected behavior

Results from `validation_step` and `validation_epoch_end` should both be logged according to what is specified. The `.log(..., logger, on_step, on_epoch)` method provides enough granularity to specify how each metric should be processed for both validation methods.

I have tried using `return [result, outputs]` in `validation_epoch_end` to work around this, but the metrics in the EvalResult are assumed to be reduced to scalars if `validation_epoch_end` is overwritten and are therefore not handled correctly.

### Environment

 - PyTorch-Lightning Version: 0.9.0
 - PyTorch Version: 1.6.0
 - OS: linux
 - How you installed PyTorch: pip
 - Python version: 3.6.9

### PR DESCRIPTION
**Pull request by gerardsn**
## What does this PR do?
Allows logging of metrics in both `validation_step` and `validation_epoch_end`. (Works for `test_step` & `test_epoch_end` as well.)
By collating `validation_step` metrics and reducing metrics with `on_epoch` flag at the start of `evaluate_loop.__run_eval_epoch_end` it is possible to continue logging metrics onto the resulting EvalResult object during `validation_epoch_end`.
Behaviour doesn't change when `using_result_obj==False`

**Current behaviour**:
```
if overridden (`validation_epoch_end`) 
    collate results and call user specified epoch end method.
else 
    collate and reduce metrics as specified in `validation_step`.
```
The user needs to create a new EvalResult object in `validation_epoch_end` since no reduction will take place and continued use of the input EvalResult object with fail when the logger attempts to write non-scalar values. As a result, all metrics (including callback_metrics) that are only defined in during `valdiation_step` are lost or need to be manually reduced and logged to the new EvalResult obj by the user when `validation_epoch_end` is overridden.

**New behaviour**:
```
collate and reduce metrics as specified in `validation_step`
if overridden (`validation_epoch_end`) 
    call user defined method
```
By reducing the results before the call to `validation_epoch_end` it is possible to continue logging onto the now collated and reduced (for step metrics with flag `on_epoch=True`) EvalResult object. 
The only change compared to the current workflow requires setting the `result.log(..., on_epoch=False)` in the `validation_step` if the non-reduced metrics are required for processing `on_epoch`. However, I think the new behaviour better matches what is intuitively expected based on flag names.

Fixes #3215 

# Before submitting

- [x] Was this discussed/~approved~ via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by gerardsn** [comment]
I am unable to pinpoint to exact issue with the tests. On my system the sanity check fails, but the actual fit works just fine and returns the correct values. Some help would be highly appreciated.

**Comment by gerardsn** [comment]
this solution is too simple and causes train/validation behaviour to diverge

**Comment by edenlightning** [comment]
@Borda mind helping out with tests?

**Comment by Borda** [comment]
@gerardsn mind resolve conflicts by rebasing on master, pls?

**Comment by gerardsn** [comment]
@Borda I changed the proposed solution for this PR.

key functions in `Result` object:
- `.gather()` -> recursively gather results
- `.reduce_on_epoch_end()` -> first lines where equivalent to .gather(), then reduce reduce results
- `.reduce_across_time()` -> same as above (changed in prep for aligning 'training_epoch_end' behaviour)

Both reduce functions now expect a gathered Result obj as input. (Maybe change from class to static method?)
Batch_sizes are now gathered as well and stored in the updated meta field.

```
This allowed changing workflow on evaluation_epoch_end from
- if epoch_end overriden: gather results and require user to reduce them.
- if epoch_end not overridden: auto reduce results.

to new behaviour:
1a. gather results
1b. optionally let user manipulate results
2.  reduce results
```

I've made some preparations to apply the same strategy to the training loop. I got stuck on the gather/reduction over time and could use some help to pull the gathering bits out of the `__auto_reduce_results_on_epoch_end()` in the logger_connector.

**Comment by edenlightning** [comment]
Hey @gerardsn! Thank you for your contribution.
We recently made a lot of changes to our logging API to make it simpler: https://pytorch-lightning.readthedocs.io/en/stable/logging.html

If this PR is still needed, please reopen and make the required changes on the new API.
